pkg/payload/task_graph_test.go

```diff
@@ -233,6 +233,143 @@ func TestByNumberAndComponent(t *testing.T) {
 	}
 }
 
+func TestFlattenByNumberAndComponent(t *testing.T) {
+	tasks := func(names ...string) []*Task {
+		var arr []*Task
+		for _, name := range names {
+			arr = append(arr, &Task{Manifest: &lib.Manifest{OriginalFilename: name}})
+		}
+		return arr
+	}
+	tests := []struct {
+		name  string
+		tasks []*Task
+		want  [][]*TaskNode
+	}{
+		{
+			name:  "empty tasks",
+			tasks: tasks(),
+			want:  nil,
+		},
+		{
+			name:  "no grouping possible",
+			tasks: tasks("a"),
+			want:  nil,
+		},
+		{
+			name:  "no recognizable groups",
+			tasks: tasks("a", "b", "c"),
+			want: [][]*TaskNode{
+				{
+					&TaskNode{Tasks: tasks("a")},
+					&TaskNode{Tasks: tasks("b")},
+					&TaskNode{Tasks: tasks("c")},
+				},
+			},
+		},
+		{
+			name:  "single grouped item",
+			tasks: tasks("0000_01_x-y-z_file1"),
+			want:  nil,
+		},
+		{
+			name:  "multiple grouped items in single node",
+			tasks: tasks("0000_01_x-y-z_file1", "0000_01_x-y-z_file2"),
+			want: [][]*TaskNode{
+				{
+					&TaskNode{Tasks: tasks("0000_01_x-y-z_file1", "0000_01_x-y-z_file2")},
+				},
+			},
+		},
+		{
+			tasks: tasks("a", "0000_01_x-y-z_file1", "c"),
+			want: [][]*TaskNode{
+				{
+					&TaskNode{Tasks: tasks("a")},
+					&TaskNode{Tasks: tasks("0000_01_x-y-z_file1")},
+					&TaskNode{Tasks: tasks("c")},
+				},
+			},
+		},
+		{
+			tasks: tasks("0000_01_x-y-z_file1", "0000_01_x-y-z_file2"),
```

Comment:
this is same as https://github.com/openshift/cluster-version-operator/pull/136/commits/fee2d0687137545ca8f8e84e91ec7b2429887c02#diff-6c3be3194acbe263715006cdee7e9442R276 `multiple grouped items in single node`